### ISSUE TITLE
Empty context for some statements when tracking "what test run this line"

### ISSUE THREAD

**Issue by davidedelvento** [coveragepy/coveragepy#974]
**Describe the bug**
Function and class definitions, as well as methods definitions and constructors come with an `(empty)` context rather than the super useful list of tests which ran the line. This is important for poorly covered code, which is the one in which coverage.py is most useful

**To Reproduce**
1. What version of Python are you using?
I think this is irrelevant, but since you ask `3.7.6`
1. What version of coverage.py are you using?
This is with coverage `5.0.4`

Running with the setting

```
[run]
dynamic_context = test_function
```
in the configuration file, against any project, and generating the report with `coverage html --show-context` generates the contexts, but only for the actual content of functions (classes, etc). The lines which start with `def` or `class` are reported with `(empty)` context. This is irrelevant for code which is well tested, since one can infer the information from the (subsequent) lines inside the function (class, etc). But for other codes which are not exercised by the test the needed information is missing.

**Expected behavior**
Context reported also for functions and class definitions, as well methods as constructors.

**Comment by efahl** [comment]
The vast majority of def and class statements are executed once at the top level when the script is run by Python for the first time, long before any test function is called.   So unless these defs and classes appear nested inside a function, they will never be executed by a test function and hence the context should remain empty (again for emphasis, with the exception of those infrequent cases where you're defining closures or what have you).

Are you trying to find out what tests import the various modules?  That's basically what appears in the context for these lines when you use the "--context=test_name" on the "coverage run" command line, but that's outside the "dynamic_context" protocol.

**Comment by davidedelvento** [comment]
Thanks, I understand the limitation now.

What I was trying to accomplish is to figure out why those definitions were reported as tested but the body of the classes/functions etc didn't. I thought a coworker wrote a partial and poorly written test. Instead, the reason was (as you said) that some other test code imported a module-under-test which (in turn) imported other modules which imported the modules which were untested. Yet, even if untested, the definitions had been executed and were counted as tested in the coverage report.

As I said, I understand the limitation now. Perhaps if you could easily remove these lines from the list of "tested" code when they actually aren't, that could be less confusing to people just starting with the project. If the logic for doing that without breaking other things is too complicated, I'm ok with just closing this issue. In such a case, I suggest adding something into the documentation about it, since it's weird to see an untested module being reported as (at least partially) tested.

**Comment by nedbat** [comment]
I understand the confusion.

But removing def and class lines wouldn't make things better. Now your coverage numbers would be lower, and looking through the reports for unexecuted lines, you'd see all the def and class lines and wonder, OK, how can I get them covered?

I don't see a useful way to change the coverage.py code.  But perhaps there's something we could do in the docs to make this clearer? Do you have any ideas?

**Comment by davidedelvento** [comment]
Yeah, I was not suggesting to remove def and class lines exactly for the reason you mentioned. That is why I wrote *If the logic for doing that without breaking other things is too complicated*, by which I meant: 

```
if only the definition (but not the body) of something is reported as covered:
    it should be changed to uncovered
else:
    leave it as it is
```

As I wrote, this may be too complicated to make it right. 

If that is the case, in the docs I would add a FAQ like:

**Q: Why do the def lines of functions (or classes) show as executed, but the bodies do not?**

which mirrors the reverse case already covered. More or less what you said in your answer here would suffice in the FAQ too.

Moreover, I would also add a sentence (perhaps even as footnote, if you prefer) in the section https://coverage.readthedocs.io/en/coverage-5.0.4/cmd.html#coverage-summary saying something like

> Note that coverage is measured by checking if a line [is executed](http://link.to.more.documentation) and therefore function and class **definitions** will be counted as covered even if the containing module is just imported, but not actually exercised by a test. See the FAQ (link) for details
